lib/vtls/openssl.c

```diff
@@ -1119,6 +1124,10 @@ static int x509_name_oneline(X509_NAME *a, char *buf, size_t size)
  */
 static int Curl_ossl_init(void)
 {
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) &&  \
+  !defined(LIBRESSL_VERSION_NUMBER)
+  OPENSSL_init_ssl(INITFLAGS, NULL);
```

Comment:
There may be things we are currently doing here that are not done by [OPENSSL_init_ssl](https://www.openssl.org/docs/man1.1.0/man3/OPENSSL_init_crypto.html). We use ENGINE_load_builtin_engines and so the equivalent of that might be OPENSSL_INIT_ENGINE_ALL_BUILTIN. They document which flags are ignored for subsequent calls (if we're not the first library to call) and that's not one of them. Ok fine. But what if another library first initializes with OPENSSL_INIT_NO_ADD_ALL_CIPHERS, which is ignored for subsequent calls, then we would not have those ciphers available if I understand this right.

Reply:
Good points. But I don't know what we can do about those doubts and unanswered questions. It seems to me that `OPENSSL_init_ssl` is now the mandated and blessed way to init the library and I would imagine that all other ways also risk getting some init glitches.

Reply:
I looked in the code for openssl 1.1.1 STABLE and ENGINE_load_builtin_engines  is just a call to OPENSSL_init_crypto(OPENSSL_INIT_ENGINE_ALL_BUILTIN, NULL) so maybe that flag should be added to curl's call ?

The openssl command's main() calls OPENSSL_init_ssl(OPENSSL_INIT_ENGINE_ALL_BUILTIN | OPENSSL_INIT_LOAD_CONFIG, NULL).

On the other hand it is probably a bad idea to call ENGINE_load_builtin_engines on openssl >= 1.1.0 as that would probably 'use up' the one chance yo seem to have to call OPENSSL_init_ssl / OPENSSL_init_crypto.

Reply:
Good find. I've added `OPENSSL_INIT_ENGINE_ALL_BUILTIN` to the bitmask now.